python_modules/dagster/dagster/_core/decorator_utils.py

```diff
@@ -128,3 +147,151 @@ def has_at_least_one_parameter(
     fn: Union[Callable[Concatenate[T, P], R], Callable[P, R]],
 ) -> TypeGuard[Callable[Concatenate[T, P], R]]:
     return len(get_function_params(fn)) >= 1
+
+
+def get_decorator_target(obj: Decoratable) -> Callable:
+    """Given a callable to be decorated, return the underlying function.
+
+    - If a function is passed, return it.
+    - If a NamedTuple class is passed, return its __new__ method.
+    - If any other class is passed, return its __init__ method.
+    - If a property is passed, return its `fget`.
+    - If a classmethod or staticmethod is passed, return its `__func__`.
+    - If a Dagster resource is passed, return its `resource_fn`.
+    - If a Dagster op is possed, return its `compute_fn`.
+    """
+    if isinstance(obj, property):
+        assert obj.fget, "Cannot decorate property without getter"
+        return obj.fget
+    elif isinstance(obj, (classmethod, staticmethod)):
+        return obj.__func__
+    elif isinstance(obj, type):
+        constructor_name = "__new__" if issubclass(obj, tuple) else "__init__"
+        return getattr(obj, constructor_name)
+    elif is_resource_def(obj):
+        return obj._resource_fn  # noqa: SLF001
+    elif is_op_def(obj):
+        if is_decorated_op_function(obj.compute_fn):
+            return obj.compute_fn.decorated_fn
+        else:
+            return obj.compute_fn  # type: ignore  # (strict type guard)
+    else:
+        return obj
+
+
+def apply_pre_call_decorator(obj: T_Decoratable, pre_call_fn: Callable[[], None]) -> T_Decoratable:
+    target = get_decorator_target(obj)
+    new_fn = _wrap_with_pre_call_fn(target, pre_call_fn)
+    return _update_decoratable(obj, new_fn)
+
+
+def _wrap_with_pre_call_fn(
+    fn: T_Callable,
+    pre_call_fn: Callable[[], None],
+) -> T_Callable:
+    @functools.wraps(fn)
+    def new_fn(*args, **kwargs):
```

Comment:
does `new_fn` show up in stack traces? may want a slightly more useful name then - its nice to be able to quickly at a glance understand what some random shit in the calls stack is there for 

`wrapped_with_pre_call_fn` or something maybe

Reply:
yes it does, good call